OpenRA.Mods.Common/Traits/Conditions/GrantChargeDrainCondition.cs

```diff
@@ -0,0 +1,223 @@
+
+		public override object Create(ActorInitializer init) { return new GrantChargeDrainCondition(init.Self, this); }
+	}
+
+	public enum TimedDeployState { Charging, Ready, Active, Deploying, Undeploying }
+
+	public class GrantChargeDrainCondition : PausableConditionalTrait<GrantChargeDrainConditionInfo>,
+		IResolveOrder, IIssueOrder, ISelectionBar, IOrderVoice, ISync, ITick, IIssueDeployOrder
+	{
+		readonly Actor self;
+		readonly bool canTurn;
+		int deployedToken = ConditionManager.InvalidConditionToken;
+		int deployingToken = ConditionManager.InvalidConditionToken;
+
+		ConditionManager manager;
+
+		[Sync]
+		int ticks;
+
+		TimedDeployState deployState;
+
+		public GrantChargeDrainCondition(Actor self, GrantChargeDrainConditionInfo info)
+			: base(info)
+		{
+			this.self = self;
+			canTurn = self.Info.HasTraitInfo<IFacingInfo>();
+		}
+
+		protected override void Created(Actor self)
+		{
+			manager = self.Trait<ConditionManager>();
+
+			if (Info.StartsFullyCharged)
+			{
+				ticks = Info.DrainTicks;
+				deployState = TimedDeployState.Ready;
+			}
+			else
+			{
+				ticks = Info.ChargeTicks;
+				deployState = TimedDeployState.Charging;
+			}
+
+			base.Created(self);
+		}
+
+		Order IIssueDeployOrder.IssueDeployOrder(Actor self, bool queued)
+		{
+			return new Order("GrantChargeDrainCondition", self, queued);
+		}
+
+		bool IIssueDeployOrder.CanIssueDeployOrder(Actor self, bool queued) { return !IsTraitPaused && !IsTraitDisabled; }
+
+		IEnumerable<IOrderTargeter> IIssueOrder.Orders
+		{
+			get
+			{
+				if (!IsTraitDisabled)
+					yield return new DeployOrderTargeter("GrantChargeDrainCondition", 5,
+						() => IsCursorBlocked() ? Info.DeployBlockedCursor : Info.DeployCursor);
+			}
+		}
+
+		Order IIssueOrder.IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID == "GrantChargeDrainCondition")
+				return new Order(order.OrderID, self, queued);
+
+			return null;
+		}
+
+		void IResolveOrder.ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString != "GrantChargeDrainCondition")
+				return;
+
+			if (!order.Queued)
+				self.CancelActivity();
+
+			if (deployState != TimedDeployState.Active)
+				self.QueueActivity(new CallFunc(Deploy));
+			else
+				self.QueueActivity(new CallFunc(RevokeDeploy));
+		}
+
+		bool IsCursorBlocked()
+		{
+			return IsTraitPaused;
+		}
+
+		string IOrderVoice.VoicePhraseForOrder(Actor self, Order order)
+		{
+			return order.OrderString == "GrantChargeDrainCondition" ? Info.Voice : null;
+		}
+
+		void Deploy()
+		{
+			deployState = TimedDeployState.Deploying;
+
+			OnDeployCompleted();
+		}
+
+		void OnDeployCompleted()
+		{
+			if (deployedToken == ConditionManager.InvalidConditionToken)
+				deployedToken = manager.GrantCondition(self, Info.DeployedCondition);
+
+			if (deployingToken != ConditionManager.InvalidConditionToken)
+				deployingToken = manager.RevokeCondition(self, deployingToken);
```

Comment:
This code will never run since `deployingToken` will always be `ConditionManager.InvalidConditionToken`.